articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+const products = {
+  '1': {
+    productName: 'xxxx',
+    price: 0
+  },
+  '2': {
+    productName: 'xxxx',
+    price: 0
+  }
+} as unknown as Record<string, ProductData>
+
+function makeRequestClosure() {
+  let count = 0
+
+  return async function (productId: string): Promise<ProductData> {
+    const product = products[productId]
+  
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        if (product) {
+          
+          if (count > 0) {
+            product.productReference = `${productId}-xxxx`
+          }
+
+          resolve(product)
+          count++
+        } else {
+          reject(new Error('It was not product found'))
+        }
+
+      }, 500)
+    })
+  }
+}
+
+const makeRequest = makeRequestClosure()
+
+const productDetails: Record<string, Record<string, string>> = {
+  '1-xxxx': {
+    care: 'Do not wash in washing machine'
+  },
+  '2-xxxx': {
+    care: 'Do not wash in washing machine'
+  }
+}
+```
+
+We just created the makeRequestClosure with a variable count initialized in 0 and it will increment its value every time a product is found. If count is greater than 0 the productReference is added to the product and the promise is inmediately resolved (If you want to dive deep into closure please at this [link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) or let us know if you want us to do an article about it).
+
+Done that, we will create a new component called ErrorFallback and this one will receive some props passed by ErrorBoundary.
+
+```tsx
+function ErrorFallback ({ error }: ErrorFallbackProps) {
+  const { resetBoundary } = useErrorBoundary()
+
+  return (
+    <div>
+      <p>There was an error: <strong>{error.message}</strong></p>
+      <button onClick={resetBoundary}>Try Again</button>
+    </div>
+  )
+}
+```
+
+We are showing the exact thrown error message but if you would rather show another one you could do it. We are also executing a customHook to get a reset function and this is connected with a function that we can pass directly in the component ErrorBoundary like this.
+
+```tsx
+<ErrorBoundary FallbackComponent={ErrorFallback} onReset={getProductData}>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+The function getProductData is the result of refactoring the useEffect:
+
+```tsx
+const getProductData = useCallback(() => {
+  if (!productId) {
+    return
+  }
+
+  setStatus('pending')
+  makeRequest(productId)
+    .then((data) => {
+      setStatus('resolved')
+      setProductData(data)
+    })
+    .catch((error: Error) => {
+      setStatus('rejected')
+      setError(error.message)
+    })
+}, [productId])
+
+React.useEffect(() => {
+  getProductData()
+}, [getProductData])
+```
+
+We isolated the logic of making the request and updating the status in a function so we can re use it inside the useEffect and outside as we do passing it as prop. By the way, the function is wrapped in an useCallback to be only updated in case productId changed.
```

Comment:
```suggestion
We isolated the logic of making the request and updating the status in a function so we can re-use it inside the useEffect and outside as we do passing it as prop. By the way, the function is wrapped in an useCallback to be only updated in case productId changed.
```